### PR TITLE
[spirv] Add verifier for lowering configurations

### PR DESCRIPTION
**Pull request by yzhang93**
The verification logic is following the constraints in #10500.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVLowerExecutableTargetPass.cpp [hunk 1]
```diff
@@ -58,6 +58,45 @@ class SPIRVLowerExecutableTargetPass
...
+  return failure(walkResult.wasInterrupted());
+}
+
+static LogicalResult verifyEntryPoint(
+    ModuleOp moduleOp, IREE::Codegen::TranslationInfoAttr translationInfo,
+    IREE::HAL::ExecutableExportOp exportOp) {
+  Optional<mlir::ArrayAttr> workgroupSizeAttr = exportOp.getWorkgroupSize();
+
+  if (workgroupSizeAttr->size() != 3) {
+    return moduleOp.emitError("expected workgroup size as three for SPIRV pipelines");
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/SPIRVLowerExecutableTargetPass.cpp:81]: ... workgroup size to have three dimensions for SPIR-V ...

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  const int maxSharedMemory = resourceLimits.getMaxComputeSharedMemorySize();
+  const int maxWorkGroupInvocations = resourceLimits.getMaxComputeWorkgroupInvocations();
+  auto maxWorkGroupSizeAttr = resourceLimits.getMaxComputeWorkgroupSize();
+  SmallVector<int64_t, 3> maxWorkGroupSize;
+  for (auto attr : llvm::enumerate(maxWorkGroupSizeAttr)) {
+    maxWorkGroupSize[attr.index()] = attr.value().cast<IntegerAttr>().getInt();
+  }
+
+  // Verify each dimension of workgroupSize should be power of two
+  if ((workgroupSize[0] != 1 && workgroupSize[0] % 2 != 0) ||
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:51]: We can use `llvm::isPowerOf2_64` directly.

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      (workgroupSize[1] != 1 && workgroupSize[1] % 2 != 0) ||
+      (workgroupSize[2] != 1 && workgroupSize[2] % 2 != 0)) {
+    return op->emitOpError("expected workgroup size to be power of two");
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:54]: expected each workgroup size dimension to be ...

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  }
+
+  // Verify each dimension of workgroup size should not exceed the corresponding
+  // limit of maxWorkGroupSize
+  if (workgroupSize[0] > maxWorkGroupSize[0] ||
+      workgroupSize[1] > maxWorkGroupSize[1] ||
+      workgroupSize[2] > maxWorkGroupSize[2]) {
+    return op->emitOpError("expected each dim of workgroup size not exceed ")
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:62]: expected workgroup size dimensions not exceeding ...

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+           << "[" << maxWorkGroupSize[0] << ", " << maxWorkGroupSize[1]
+           << ", " << maxWorkGroupSize[2] << "]";
+  }
+
+  // Verify the total workgroup size should not exceed maxWorkGroupInvocations
+  int64_t totalWorkgroupSize =
+      workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
+  if (totalWorkgroupSize > maxWorkGroupInvocations) {
+    return op->emitOpError("expected total workgroup size to be <= ")
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:71]: expected total invocation count in workgroup to be ...

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+      firstLevelTileSizes.erase(firstLevelTileSizes.begin());
+      secondLevelTileSizes.erase(secondLevelTileSizes.begin());
+      thirdLevelTileSizes.erase(thirdLevelTileSizes.begin());
+    }
+
+    // Verify the tile size divides the matmul inputs A [M x K] & B [K x N]
+    if (lhsShape[0] % firstLevelTileSizes[0] != 0 ||
+        lhsShape[1] % thirdLevelTileSizes[2] != 0) {
+      return op->emitOpError(
+                 "lhsShape doesn't factor into first level tile size");
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:117]: LHS shape is indivisible by first level tile size

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    }
+    if (rhsShape[0] % thirdLevelTileSizes[2] != 0 ||
+        rhsShape[1] % firstLevelTileSizes[1] != 0) {
+      return op->emitOpError(
+                 "rhsShape doesn't factor into first level tile size");
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:122]: same here

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    unsigned totalSharedMemSizeBytes =
+        (firstLevelTileSizes[0] * thirdLevelTileSizes[2] +
+         firstLevelTileSizes[1] * thirdLevelTileSizes[2]) * bytesSize;
+
+    if (totalSharedMemSizeBytes > maxSharedMemory) {
+      return op->emitOpError("expected shared memory usage <= ")
+             << maxSharedMemory << ", got " << totalSharedMemSizeBytes;
+    }
+
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:144]: Use an early `return success()` here?

**Review by antiagainst — CHANGES_REQUESTED** [review]
Awesome, thanks for adding these! I Generally looks good; just some nits.

**Comment by yzhang93** [comment]
Thanks for your review and comments! I've modified codes accordingly.

**Review by antiagainst — APPROVED** [review]
Awesome. Just two final nits. :)

**Comment by yzhang93** [comment]
I just realized many tests in https://github.com/iree-org/iree/tree/main/compiler/src/iree/compiler/Codegen/SPIRV/test and tests/e2e/models/check_vulkan-spirv_vulkan_mobilenetv3_fake_weights.mlir_module.vmfb were failed because of the verifier constraints. Many of them were failed because of "error: expected total workgroup size to be multiple of xx". Shall I modify the configs in these files? Or shall we relax some constraints? For example, if our input matmul dimensions are odd numbers, with the current constraints, it's hard to make the workgroup size multiple of 16/32/64.

**Comment by antiagainst** [comment]
Looking into this I think it's hard to verify `SPIRVVectorize` directly, which is used by many things. So I sent out two patches #10735 & #10736 to clean up pass pipelines. After that we will have separate identifiable pipelines for matmul and convolution. After landing them this patch should be adjusted to verify `SPIRVMatmulPromoteVectorize` for matmul and `SPIRVConvDirectVectorize` for convolution respectively.

(BTW I also pushed a commit 1890f1b to this pull request to address a few issues when I was touching it. Please make sure you `git pull` to get them.)

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by yzhang93** [comment]
@antiagainst I've modified the codes accordingly. Could you please review this again when possible?

**Review by antiagainst — APPROVED** [review]
LGTM! Could you fix the formatting?

**Comment by yzhang93** [comment]
> LGTM! Could you fix the formatting?

Oh, what is the correct format? I already did git clang-format HEAD~1 and force pushed to my branch.

**Comment by antiagainst** [comment]
Lint tests is still failing here: https://github.com/iree-org/iree/actions/runs/3434414360/jobs/5736372447. Nvm, I'll fix that and land it. :)

**Comment by yzhang93** [comment]
> Lint tests is still failing here: https://github.com/iree-org/iree/actions/runs/3434414360/jobs/5736372447. Nvm, I'll fix that and land it. :)

Awesome! Thank you!

**Pull request merged**